pkg/controller/usersignup/usersignup_controller.go

```diff
@@ -240,8 +288,8 @@ func (r *ReconcileUserSignup) provisionMasterUserRecord(userSignup *toolchainv1a
 			"Error creating MasterUserRecord")
 	}
 
-	logger.Info("#### Created MasterUserRecord", "Name", mur.Name, "TargetCluster", targetCluster)
-	return nil
+	logger.Info("Created MasterUserRecord", "Name", mur.Name, "TargetCluster", targetCluster)
+	return r.updateStatusCompliantUsername(userSignup)
```

Comment:
Oh... if we fail here (during status update) then we are screwed. It means the MUR already exists but we don’t know about that and we will create another one with regenerated compliant username in the next reconcile cycle. And in case of some constant error (due to some bug for example) we would end up with a 100/1000 MURs created for the single UserSignup.

So we have to use userID label in MUR as I mentioned in my comment above. UserID is always available to us and it doesn’t change.

Reply:
That won't happen.. check the logic in `generateCompliantUsername()` again - if a MUR is found with the same UserID as the UserSignup, then that one will be used.

Reply:
Yeah I missed that but unfortunately it won't solve the problem.
We may end up with multiple usersignups with the same compliant username. And as soon the first usersignup is approved and creates the corresponding MUR that MUR will be recognized ad the MUR for all these signups - https://github.com/codeready-toolchain/host-operator/pull/93/files#diff-5b4b18a0fc2e0d708e4e6334cdb6bc69R158
We could check user ID there too and trigger username regeneration but IMO the logic will be clearer and less error-prone if we rely on the userID label in MUR and generate the username only when the signup is approved. Right before creating a MUR.

Reply:
I checked my business logic again and I'm pretty sure I got it correct.  I don't see how it is possible for multiple usersignups to have the same compliant username, the code shouldn't allow it.  Also, the code I have [here](https://github.com/codeready-toolchain/host-operator/pull/93/files#diff-5b4b18a0fc2e0d708e4e6334cdb6bc69R134) is (I'm pretty sure) correct too.  The `generateCompliantUsername()` call will ensure that if the MUR for that UserID already exists then the `compliantUsername` will be set to that value.  This takes care of the situation where the MUR exists already but the `UserSignup.Status.CompliantUsername` isn't set yet (for whatever strange reason).

Reply:
This is actually really complex business logic, and it took me a few hours to work out the exact process to handle all possible scenarios.  I think it might be worth having a bluejeans call to go over the details here.

Reply:
Yeah right. I missed that status set only if the MUR with such compliant username **and** userID found. And not when the generated name available. But unless I'm missing something again it causes another issue - https://github.com/codeready-toolchain/host-operator/pull/93#discussion_r341835246
It can be fixed by passing the generated name to MUR provisioning method instead of using the status (which is not set at that point) but I still have a feeling that using UserID label in MUR would simplify the logic.

Reply:
Good catch here, I did miss this one.  It should now handle all scenarios - I'll work on determining them all and get them covered in the e2e tests.